Add scalar.conj() and update backward formulas for add and sub

anjali411 opened the pull request:
1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46596 Add scalar.conj() and update backward formulas for add and sub**

Differential Revision: [D24529227](https://our.internmc.facebook.com/intern/diff/D24529227)

anjali411 committed:
Add scalar.conj() and update backward formulas for add and sub

[ghstack-poisoned]

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).




[ghstack-poisoned]

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).




[ghstack-poisoned]

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).




[ghstack-poisoned]

ezyang commented on c10/core/Scalar.cpp:
```diff
@@ -13,4 +13,14 @@ Scalar Scalar::operator-() const {
+  if (isComplex()) {
+    return Scalar(std::conj(v.z));
+  } else if (isFloatingPoint()) {
+    return Scalar(v.d);
```
can't you just return `*this` here?

ezyang reviewed (approved):
At some point when you're not in a rush, it would be good to use @robieta's tools for instruction counts to assess what the non-complex overheads of the new complex idioms and conjugations are. It will be a little tricky to setup since you have to go through autograd, but we might be able to setup a microbenchmark where we just call the microbenchmark directly. That will help us understand how urgent extra optimizing here is (which I know @ngimel is nervous about the conj additions here.)

mruberry reviewed (approved):
Test changes look good. Test failure doesn't appear to be related.

gchanan commented on aten/src/ATen/test/scalar_test.cpp:
```diff
@@ -128,3 +128,13 @@ TEST(TestScalar, TestScalar) {
+TEST(TestScalar, TestConj) {
+  Scalar int_scalar = 257;
+  Scalar float_scalar = 3.0;
+  Scalar complex_scalar = c10::complex<double>(2, 3);
```
nit: probably good to use non .0 values just to be sure you haven't accidentally done something like cast to int first.

anjali411 replied:
I am not sure I follow ...

ezyang replied:
Greg is saying that if there was a a bug where the Scalar constructor for complex numbers accidentally represented the complex number as integers, you wouldn't catch it with this test. Need fractional value to see it.

anjali411 replied:
oh ok

gchanan commented on tools/autograd/gen_variable_type.py:
```diff
@@ -165,7 +165,7 @@
     'unbind', 'split', 'split_with_sizes', 'unsafe_split', 'split_with_sizes_backward',
     'dot', 'vdot', 'cholesky', 'triangular_solve', 'mm', '_unsafe_view', 'mv', 'ger',
-    'bmm', 'diagonal'
+    'bmm', 'diagonal', 'nonzero'
```
should this be a different PR?  what does this have to do with the other changes?

anjali411 replied:
yeah it is an unrelated change that's true

gchanan commented on tools/autograd/derivatives.yaml:
```diff
@@ -995,11 +995,11 @@
 - name: sub.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
-  self: grad
-  other: -grad * alpha
+  self: handle_r_to_c(self.scalar_type(), grad)
```
why do we need this?  This seems pretty similar to https://github.com/pytorch/pytorch/blob/e02a3e190ec79b2ac9a4fbc3428473503889c690/torch/csrc/autograd/engine.cpp#L611 -- is it because `to` doesn't work c->r?

anjali411 replied:
.to works for `C->R`. currently `handle_r_to_c` returns a (real) view of the final gradient value in case the gradient is complex and the input is not.

anjali411 commented:
> At some point when you're not in a rush, it would be good to use @robieta's tools for instruction counts to assess what the non-complex overheads of the new complex idioms and conjugations are. It will be a little tricky to setup since you have to go through autograd, but we might be able to setup a microbenchmark where we just call the microbenchmark directly. That will help us understand how urgent extra optimizing here is (which I know @ngimel is nervous about the conj additions here.)

I think that's a great idea. is there a doc to help with the setup?

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).




[ghstack-poisoned]

ezyang commented:
Hopefully Taylor's talk today should help? He's also been working on written docs as well.

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).


Differential Revision: [D24529227](https://our.internmc.facebook.com/intern/diff/D24529227)

[ghstack-poisoned]

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).


Differential Revision: [D24529227](https://our.internmc.facebook.com/intern/diff/D24529227)

[ghstack-poisoned]

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).


Differential Revision: [D24529227](https://our.internmc.facebook.com/intern/diff/D24529227)

[ghstack-poisoned]

anjali411 committed:
Update on "Add scalar.conj() and update backward formulas for add and sub"

1. Added `conj` method for scalar similar to numpy.
2. Updates backward formulas for add and sub to work correctly for R -> C cases and for the case when alpha is complex.
3. Enabled complex backward for nonzero (no formula update needed).


Differential Revision: [D24529227](https://our.internmc.facebook.com/intern/diff/D24529227)

[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -538,6 +538,7 @@ def method_tests():
         ('add', (), ((S, S, S),), 'scalar_broadcast_lhs', (True,)),
         ('add', (S, S, S), (3.14,), 'constant', (True,)),
         ('add', (), (3.14,), 'scalar_constant', (True,)),
+        ('add', (S, S, S), (3.14j,), 'complex_scalar_constant', (True,)),
```
Does this actually calls into the scalar overload? I think that the python binding is converting these to Tensors all the time.

You can double check this by doing: 
```python
import torch

a = torch.rand(10, requires_grad=True)
print((a + a).grad_fn)
print((a + 1.).grad_fn)
```
That will print 
```
<AddBackward0 object at 0x7fc1c1446150>
<AddBackward0 object at 0x7fc1c2f14f50>
```

The important part is the `0` here that means that both are actually using the 0th overload (the one for Tensor inputs).
